MAINT: Rewrite Floyd algorithm

thrasibule opened the pull request:
- use a simple hash table with linear probing
- adjust heuristic

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

thrasibule committed:
Rewrite Floyd algorithm

- use a simple hash table with linear probing
- adjust heuristic

thrasibule commented:
This is to address https://github.com/scipy/scipy/issues/9699. See also #5299 To sample sparse matrices with a given density of non zero values, scipy calls choice(N, k, replace=True), where N large and k<<N. The best algorithm is Floyd's algorithm which is O(k). The current implementation wasn't used most of the time given to an incorrect heuristic (looks like it assumed Floyd algorithm is O(k**2) where it's actually O(k)).

I've also replaced the set structure in python with a simple hash table which has much lower memory usage and is quite a bit faster. Timing tests show that the new version of the Floyd sampling beats the Fisher-Yates shuffle even when k~N.

bashtage commented on numpy/random/generator.pyx:
```diff
@@ -685,26 +690,31 @@ cdef class Generator:
+                    set_size = <uint64_t>(1.2 * size_i)
+                    mask = _gen_mask(set_size)
+                    set_size = 1 + mask
+                    hash_set = <uint64_t*>malloc(sizeof(uint64_t) * set_size)
```
The gain to using malloc and free seem pretty small when compared to allocating an array with dtype uint64_t.  If you go this way need to gracefully handle the unlikely event that malloc fails (if you use an array, you get the handling for free).  There is also a question as to whether you should be using malloc/free or PyMem_Malloc/PyMem_Free, which I'm not certain about. The latter is faster under some (unknown to me) conditions.

bashtage replied:
Changes look good.

bashtage commented on numpy/random/generator.pyx:
```diff
@@ -685,26 +690,31 @@ cdef class Generator:
-                                idx_data[loc] = val = j
-                            idx_set.add(val)
-                            loc += 1
+                            val = random_interval(&self._bitgen, j)
```
You should be using `random_bounded_uint64(&self._bitgen, 0, j, 0, 0)` here which is faster for most j that aren't very close to powers of 2.

bashtage commented:
> The current implementation wasn't used most of the time given to an incorrect heuristic (looks like it assumed Floyd algorithm is O(k**2) where it's actually O(k)).

The O(k**2) comes from two features.  1. A lookup in a Python set is worst case O(n) where n is the number of elements. I don't know where the n kicks in.  If it is only if it hits O(n) well before n\approx k (in particular,  for some value where 1-n/k > c > 0 for some c > 0, then one would end up checking c*k at a cost of O(k-n), hence the O(k**2).  This plus timing lead to the improved tail sort (which is O(pop_size) since it allocates an array 1:pop_size).

bashtage commented:
> Timing tests show that the new version of the Floyd sampling beats the Fisher-Yates shuffle even when k~N.

If you have time it would be good to improve the heuristic since you are still switching for k >  .1N

bashtage commented:
If you could show that it is faster for k>=0.5N even if it slower for k > x%N for large x near 1 (and large N) then it might be possible to return the compliment (although not sure about how to order though).

bashtage commented:
How does this compare to #10124 if p is set to 1/pop_size?

thrasibule commented:
For the set complexity, my understanding is a lookup is O(1) amortized, and worst case O(n) if the  hash table behind it gets resized. But we know exactly how big the hash table needs to be, which is what I did. There is also no need to worry about collisions because we're sampling uniformly from 1..N, so the data is as well behaved as possible.
So theoretically they're both O(k), but Fisher Yates is O(n) in space and Floyd is O(k). The only reason I still have an heuristic is that when k=n, it seems hard to bit Fisher-Yates which just does n swaps, which is as fast as it gets.

thrasibule commented:
To give an idea, timing results on my machine.
For the rewritten Floyd algorithm:
```
In [5]: %timeit choice(10_000, 10_000, replace=False)        
101 µs ± 1.14 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: %timeit choice(100_000, 100_000, replace=False)    
861 µs ± 11.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit choice(1_000_000, 1_000_000, replace=False)                                                 
14.1 ms ± 314 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: %timeit choice(10_000_000, 10_000_000, replace=False)                                                 
498 ms ± 9.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit choice(100_000_000, 100_000_000, replace=False)                                             
6 s ± 53.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: %timeit choice(1_000_000_000, 100_000_000, replace=False)                                           
9.01 s ± 70.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit choice(10_000_000_000, 100_000_000, replace=False)                                          
9.99 s ± 147 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
for tail Fisher-Yates shuffle:
```
In [2]: %timeit choice(10_000, 10_000, replace=False)      
172 µs ± 1.83 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [3]: %timeit choice(100_000, 100_000, replace=False)    
1.62 ms ± 13.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: %timeit choice(1_000_000, 1_000_000, replace=False)                                         
24.6 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [5]: %timeit choice(10_000_000, 10_000_000, replace=False)                                                 
573 ms ± 13.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit choice(100_000_000, 100_000_000, replace=False)                                               
6.95 s ± 117 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
So Floyd is actually slightly faster. I'm not quite sure why. Maybe _shuffle_raw needs to handle strides, etc... which slows it down.

thrasibule committed:
use typed memoryview instead of C array

bashtage commented:
Given your timings it is tempting to use this in all cases. Plus it reduces a branch.

I can see how your specialized hash table is much faster than using a Python set.

thrasibule commented:
> How does this compare to #10124 if p is set to 1/pop_size?

This algorithm is O(n) since it needs to draw n exponentials, and then find the top k values which is also O(n) https://en.wikipedia.org/wiki/Selection_algorithm. So worse than Floyd.

thrasibule committed:
no wraparound

thrasibule committed:
fix tests

wkschwartz commented:
Can this close #5158?

bashtage commented:
@mattip This is worth serious consideration for 1.17, when you have time.  It is much easier to get this in now than to wait until after 1.17 since it will change the ordering for a given seed/bit generator.  This implementation is a far superior one to the naive version that is currently in Generator.

charris commented:
Needs rebase. Is that test correct? @bashtage Just to be sure, you think this should go in?

bashtage commented:
It is a definite improvement, and since changing the stream is still difficult (just not impossible), it is better to get it in now rather than later. 

@thrasibule could you do a simple statistical analysis of the changes to make sure that it appears to have the correct properties? You couldn't for example, sample `k` w/o replacement from arange(0,100) many times and then compute the historgram of the empirical frequency of any observation appearing.  You could also do more sophisticated tests, but a histogram is simple and would likely show any glaring issues.

thrasibule commented:
@bashtage something like that ?

```
In [1]: import numpy as np                                                     

In [2]: from numpy.random.generator import choice                              

In [3]: nsim = 10_000 
   ...: sample_size = 100_000 
   ...: for k in [100, 200, 300, 500, 1000, 5000, 10_000]: 
   ...:     counts = np.zeros(100_000) 
   ...:     for l in range(10_000): 
   ...:         counts += np.bincount(choice(100_000, k, replace=False), minlen
   ...: gth=100_000) 
   ...:     expected_var = nsim * k/sample_size * (1-k/sample_size) 
   ...:     print(counts.var(), expected_var) 
   ...:                                                                        
9.9686 9.99
20.0336 19.96
29.98722 29.91
49.85672 49.75
99.60878 99.0
478.04698 475.0
903.26496 900.0
```

thrasibule commented:
I have a couple of questions:
- should we keep the heuristic which switches to tail shuffle, or just use Floyd?
- I've disabled wraparound (cython.wraparound(False) ) in the cython block that I wrote, which makes sense since I build a power of 2 sized hash table to specifically have fast wraparound, so no need for cython to do its slow wraparound on top. I noticed the choice method is marked cython.wraparound(True), even though it doesn't look like code in that method relies on the wraparound. Should I just remote that line then?
- One difference between the tail shuffle algorithm and Floyd, is that with tail shuffle, the sample will come in random order. In particular any slice of size k' < k will be a valid sample without replacement. Whereas with Floyd, the order is arbitrary but not uniform among the permutations of size k. This is especially clear when n=k, where Floyd will always return 0..n-1 in that order. Maybe this should be mentioned in the docstring?

thrasibule committed:
prevent warnings

charris commented:
> One difference between the tail shuffle algorithm and Floyd, is that with tail shuffle, the sample will come in random order. In particular any slice of size k' < k will be a valid sample without replacement. Whereas with Floyd, the order is arbitrary but not uniform among the permutations of size k. This is especially clear when n=k, where Floyd will always return 0..n-1 in that order. Maybe this should be mentioned in the docstring?

Is this a concern? I vaguely recall that application was mentioned at some point with regard to bootstrapping(?) Or maybe that was for selection with replacement.

bashtage commented:
> * One difference between the tail shuffle algorithm and Floyd, is that with tail shuffle, the sample will come in random order. In particular any slice of size k' < k will be a valid sample without replacement. Whereas with Floyd, the order is arbitrary but not uniform among the permutations of size k. This is especially clear when n=k, where Floyd will always return 0..n-1 in that order. Maybe this should be mentioned in the docstring?

Is this specific to this implementation? I don't think that is true for the Floyd implementation in master (just checked, it is not, e.g., `g.choice(np.arange(100),100)`). This is undesirable IMO, and feels a bit like an API change since one really needs to shuffle the choice to get the same API as the current one when n=k. 

Can this be worked around if a different hash implementation was used (i.e. faster than set() but perhaps slower than your table)?

This makes switching between Floyd and tail seem stranger as well. 

> I've disabled wraparound (cython.wraparound(False) ) in the cython block that I wrote, which makes sense since I build a power of 2 sized hash table to specifically have fast wraparound, so no need for cython to do its slow wraparound on top. I noticed the choice method is marked cython.wraparound(True), even though it doesn't look like code in that method relies on the wraparound. Should I just remote that line then?

This is probably vestigial, but one has to be certain.

bashtage commented:
Just a thought -- could you store a second table that has the order information for each value selected, and then do a quick reorder at the end?

So that when n=k, the order that each value is selected would be preserved in the output.  For example, if 0 is selected 10th, then out[9] = 0, while if 0 is selected last, then out[-1] = 0.

thrasibule commented:
> Is this a concern? I vaguely recall that application was mentioned at some point with regard to bootstrapping(?) Or maybe that was for selection with replacement.

It depends what the use case is. A specific use case could be to have the sample in sorted order actually, so having it shuffled doesn't bring any benefit.

> Is this specific to this implementation? I don't think that is true for the Floyd implementation in master (just checked, it is not, e.g., g.choice(np.arange(100),100)). This is undesirable IMO, and feels a bit like an API change since one really needs to shuffle the choice to get the same API as the current one when n=k.

No the implementation in master already had the same issue. You couldn't see it because the heuristic does a tail shuffle if k=n. When k <n, with Floyd's algorithm, the indices will tend to be lower at the beginning of the sample than at the end of the sample.

I agree that the change in behavior is strange if we keep both implementations. So we can either just keep Floyd's and specify in the doc string that the sample is not a permutation. Or keep both and do a shuffle at the end of Floyd's algorithm (this will still be O(k)).

bashtage commented:
Yup - you are completely correct.  I see it now.  (I also had the wrong command above).

IMO we need a shuffle at the since any users who expected similar behavior w.r.t. order from RandomState could be very surprised.  This much have an impact on the break-even point I suspect.  Around 50%?

bashtage commented:
If you want to consider a plain Floyd with no shuffle then I think this needs to go to the mailing list since it seems to be more than a refactor.

charris commented:
@bashtage What should we do with this for 1.17?

wkschwartz commented:
I am just one user, but I am in favor of including the new algorithm _without_ a shuffle at the end in v1.17. I could really use the fastest `choice` I can get my hands on. If I need a random permutation, I’ll call choice followed by permutation.

This is the one chance to simplify the API. Generator already is _not_ a drop in replacement for RandomState because they’re not stream compatible.

If the change in API is too much to stomach, at least expose the fast path in a separate function so I can call it without the shuffle.

charris commented:
I'm going to make the 1.17.0rc1 without this, but it will be several weeks before the final release so there will be plenty of time to put this in.

seberg commented:
At least on first sight, I am -1 on not shuffling the result. I agree that in many use cases it is not necessary (i.e. data splitting where order does does not matter), but it seems very dangerous to assume users are aware of such a change and not totally unreasonable to assume that a random function outputs random values.

A new name might be best if we want it, that is also easy to add later on. `subsample` might work, although close to `sample` (which, however, we removed from the new API).

thrasibule committed:
fix tests

mattip commented:
If we add a 'ensure_shuffle=True' kwarg, that the advanced user could set to False, and add a `shuffle` inplace step at the end, does the default `True` vaule kill the performance gain?

thrasibule committed:
faster shuffle

thrasibule committed:
extra shuffle at the end and tweak heuristic

thrasibule committed:
nogil

thrasibule committed:
fix tests

mattip commented:
This is better now that it shuffles the result.

Could you add a benchmark to `benchmark/benchmarks/bench_random.py`? Something like the one below would be nice, but maybe the 1e8 case takes too long (4 secs on my computer). Is there a reason to prefer benchmarking `choice(array, len(array), replace=False)` rather than something like `choice(array, len(array) // 10, replace=False)` ?

Note that rebasing off master will require using `np.random.default_rng()` instead of `np.random.Generator()`

<details>

```
class Choice(Benchmark):
    params = [1e3, 1e6, 1e8]
    def setup(self, v):
        self.a = np.arange(v)
        self.rng = np.random.Generator()

    def time_legacy_choice(self, v):
        np.random.choice(self.a, int(v), replace=False)

    def time_choice(self, v):
        self.rng.choice(self.a, int(v), replace=False)
```

bashtage commented on numpy/random/generator.pyx:
```diff
@@ -675,36 +674,40 @@ cdef class Generator:
                 pop_size_i = pop_size
                 # This is a heuristic tuning. should be improvable
-                if pop_size_i > 200 and (size > 200 or size > (10 * pop_size // size)):
+                if pop_size_i > 10000 and (size_i > (pop_size_i // 50)):
```
How sensitive is the perf now with the improved shuffle? Here the full array is created whenever size is > 2% pop_size.  This could be memory unfriendly when pop_size is quite large (100_000_000) but size is still modest (~10%).

thrasibule replied:
Here are some timings on my machine:

original shuffle
```
In [4]: %timeit rg.choice(100_000_000, 2_000_000, replace=: False)                                            
235 ms ± 12.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
766 ms ± 9.58 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
7.06 s ± 85.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

specialized shuffle
```
In [3]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
187 ms ± 817 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
326 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
550 ms ± 7.94 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
4.89 s ± 16.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Floyd no shuffle
```
In [4]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
111 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [5]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
320 ms ± 6.46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
641 ms ± 9.64 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
5.64 s ± 90.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Floyd with shuffle
```
In [3]: %timeit rg.choice(100_000_000, 2_000_000, replace=False)                                            
169 ms ± 836 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit rg.choice(100_000_000, 5_000_000, replace=False)                                            
496 ms ± 8.94 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit rg.choice(100_000_000, 10_000_000, replace=False)                                           
1.02 s ± 8.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit rg.choice(100_000_000, 100_000_000, replace=False)                                          
11.1 s ± 28.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
I sized the heuristic so that the speed is roughly the same when it switches from shuffle to Floyd. If we don't do a shuffle at the end, then the breakeven point is probably pop_size//20.

thrasibule replied:
This is purely timing based. You raise a good point about memory, Floyd is more memory efficient indeed, but it's a tradeoff.

thrasibule commented:
@mattip I will add the benchmark.
 For what it's worth, I'm also -1 on doing an extra shuffle, since one can always shuffle the result afterwards if one needs shuffled samples. I can implement @mattip suggestion of adding an ensure_shuffle=True keyword, so that it doesn't change the api by default. It complexifies the code a little bit, because the heuristic will now depend on whether we do the shuffle or not at the end.

wkschwartz commented:
Any reason not to call the keyword argument simply `shuffle` rather than `ensure_shuffle`?

bashtage commented:
I also like the keyword argument. As for the code, you could pay a small
perf penalty and just always have Floyd, and the keyword dependent shuffle.
This also makes me happy since it avoids creating a large array when a
modest size array will suffice.

On Sun, Jul 7, 2019, 14:21 Thrasibule <notifications@github.com> wrote:

> @mattip <https://github.com/mattip> I will add the benchmark.
> For what it's worth, I'm also -1 on doing an extra shuffle, since one can
> always shuffle the result afterwards if one needs shuffled samples. I can
> implement @mattip <https://github.com/mattip> suggestion of adding an
> ensure_shuffle=True keyword, so that it doesn't change the api by default.
> It complexifies the code a little bit, because the heuristic will now
> depend on whether we do the shuffle or not at the end.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/13812?email_source=notifications&email_token=ABKTSRNBSJLUHMT2GLDS4WLP6HUWTA5CNFSM4HZ5S5M2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZLLNPQ#issuecomment-508999358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKTSRIEY4RXNYQNBLCRCBLP6HUWTANCNFSM4HZ5S5MQ>
> .
>

bashtage commented:
Also +1 for shuffle over ensure shuffle. Probably need to run by the
mailing list.

On Sun, Jul 7, 2019, 14:36 Kevin Sheppard <kevin.k.sheppard@gmail.com>
wrote:

> I also like the keyword argument. As for the code, you could pay a small
> perf penalty and just always have Floyd, and the keyword dependent shuffle.
> This also makes me happy since it avoids creating a large array when a
> modest size array will suffice.
>
> On Sun, Jul 7, 2019, 14:21 Thrasibule <notifications@github.com> wrote:
>
>> @mattip <https://github.com/mattip> I will add the benchmark.
>> For what it's worth, I'm also -1 on doing an extra shuffle, since one can
>> always shuffle the result afterwards if one needs shuffled samples. I can
>> implement @mattip <https://github.com/mattip> suggestion of adding an
>> ensure_shuffle=True keyword, so that it doesn't change the api by default.
>> It complexifies the code a little bit, because the heuristic will now
>> depend on whether we do the shuffle or not at the end.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/numpy/numpy/pull/13812?email_source=notifications&email_token=ABKTSRNBSJLUHMT2GLDS4WLP6HUWTA5CNFSM4HZ5S5M2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZLLNPQ#issuecomment-508999358>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABKTSRIEY4RXNYQNBLCRCBLP6HUWTANCNFSM4HZ5S5MQ>
>> .
>>
>

wkschwartz commented:
You’ve already done all the experiments to figure out the heuristic, which is just an extra line when you add in the branch for the keyword-dependent shuffle. Is it really that much more complex of code that you’d let performance suffer?

bashtage commented:
> You’ve already done all the experiments to figure out the heuristic, which is just an extra line when you add in the branch for the keyword-dependent shuffle. Is it really that much more complex of code that you’d let performance suffer?

I thought it was more involved, but in psuedo-code I think it would be:

```
if pop_size is small or size/pop_size is not too big or not shuffle:
       floyd
       if shuffle:
           do_shuffle
else:
      tail_shuffle
```

This said, I still think there is a case to be made for limiting the allocation of a large array when it isn't actually required, but then the tuning requires a lot of guess-work.  A better solution would probably be to add something to the notes that explains that using `shuffle=False` minimizes memory use, and if you need a shuffle, then you can always do it later. But this is also difficult to fully explain since it would only be completely accurate in some circumstances.

thrasibule commented:
I was thinking more something along these lines:
```
if shuffle:
    cutoff = 50
else:
    cutoff = 20

if pop_size is small or pop_size // size > cutoff:

       floyd

       if shuffle:

           do_shuffle

else:

      tail_shuffle
```
Seems like no solution is perfect anyway unless we expose all the knobs. I'll try to send an email to the mailing list summarizing the discussion so far. In any case, whatever gets decided will be strictly faster than before, so it's still a net win, even if not optimal in all situations.

mattip commented:
Sent the `shuffle=True` kwarg change to the [mailing list](https://mail.python.org/pipermail/numpy-discussion/2019-July/079827.html)

thrasibule committed:
add benchmark

thrasibule committed:
add shuffle keyword

mattip commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -568,7 +568,7 @@ def test_choice_nonuniform_replace(self):
         random = Generator(MT19937(self.seed))
         actual = random.choice(4, 3, replace=False)
-        desired = np.array([0, 1, 3], dtype=np.int64)
+        desired = np.array([2, 0, 3], dtype=np.int64)
```
This would be a good place to add a test that shows the use of the `shuffle` kwarg

thrasibule replied:
added a test, thanks.

mattip commented on numpy/random/generator.pyx:
```diff
@@ -675,36 +676,45 @@ cdef class Generator:
+                                hash_set[loc] = j
+                                idx_data[j - pop_size_i + size_i] = j
+                        if shuffle:
+                            self._shuffle_int(size_i, 1, idx_data)
```
Is there a reason not to reuse the existing `shuffle` or `_shuffle_raw` ?

thrasibule replied:
I've added _shuffle_int as a specialized version of _shuffle_raw because it's ~30% faster for integers.

mattip commented on benchmarks/benchmarks/bench_random.py:
```diff
@@ -173,3 +173,16 @@ def time_bounded(self, bitgen, args):
+
+    def setup(self, v):
+        self.a = np.arange(v)
+        self.rng = np.random.Generator()
```
This is failing, either needs to be `Generator(PCG64())` or `default_gen()`

godaygo replied:
Wasn't `default_gen` renamed to `default_rng`?

mattip replied:
yes, thanks

thrasibule replied:
fixed, thanks.

thrasibule committed:
test shuffle keyword

thrasibule commented:
Thanks @mattip for the post on the mailing list. I've tried to address all the suggestions so far.

I've also experimented with using a bitset instead of a hash table. It makes sense to use it from a memory perspective when k>=N/64. It's also slightly faster because testing for an element is just a bitshift and an and operation. I've written the code in a different branch, see [here](https://github.com/thrasibule/numpy/commit/20c2ab015f66f8aaea38b3e245aeb26a95b44131). I've refactored the logic of the Floyd algorithm into 2 functions which I've added as static methods to the Generator class. I'm not sure it's the best place to put them. Maybe generator.pxd would be better. Anyway here are some timings: it's faster in the shuffle=False, k>=N/50 range.

N,k| HashSet, shuffle=False  | BitSet, shuffle=False | HashSet, shuffle=True| Bitset, shuffle=True|
---| -----------------------  | -------------------- |-----------------------|--------------------|
N=1e8, k=1e5|1.97 ms ± 12.4 µs|2.07 ms ± 13.3 µs|2.79 ms ± 19.6 µs |3.02 ms ± 35.9 µs|
N=1e8, k=1e6|54.3 ms ± 1.06 ms   |52.8 ms ± 1.71 ms |67.6 ms ± 639 µs|65.1 ms ± 1.7 ms|
N=1e8, k=2e6|125 ms ± 1.09 ms|104 ms ± 1.36 ms |175 ms ± 2.09 ms|147 ms ± 1.44 ms|
N=1e8, k=5e6|354 ms ± 6.78 ms|265 ms ± 4.75 ms|372 ms ± 2.38 ms|378 ms ± 9.21 ms|
N=1e8, k=1e7|610 ms ± 8.72 ms|507 ms ± 6.06 ms|625 ms ± 9.38 ms|629 ms ± 6.28 ms|
N=1e8, k=5e7|2.54 s ± 35.6 ms|2.09 s ± 44.1 ms|2.53 s ± 11.5 ms|2.62 s ± 13.3 ms|
N=1e8, k=1e8|4.81 s ± 77 ms|2.93 s ± 21.2 ms|4.74 s ± 65.6 ms|4.85 s ± 49.5 ms|

charris commented:
Thanks @thrasibule .
